### ISSUE TITLE
multiple tests fail when using pyOpenSSL 0.14

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7370]
|<img alt="felixonmars's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| felixonmars reported|
|-|-|
|Trac ID|trac#7370|
|Type|defect|
|Created|2014-05-18 15:24:07Z|
|Branch|https://github.com/twisted/twisted/tree/failing-tls-tests-7370|

These tests pass (or skip) when using pyOpenSSL 0.13 and fail when using pyOpenSSL 0.14.

See attachment for log.

Attachments:

* [twistederrors.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/654/654484cfcdc1f1f2902706c8c6855b95e59cdd08/0d08e18b76d6ba6d6cee26274e3d02843ebb738a.txt) (20284 bytes) - added by felixonmars on 2014-05-18 15:24:42Z - 
* [twisted_test_failures_with_pyopenssl_0.13.1.log](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/654/654484cfcdc1f1f2902706c8c6855b95e59cdd08/09504bc628cf419bdeb142be9672ddf6792fc5bd.log) (2404 bytes) - added by felixonmars on 2014-05-22 10:19:46Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7370 7370
type__defect defect
reporter__felixonmars felixonmars
priority__normal normal
milestone__None None
branch__branches_failing_tls_tests_7370 branches/failing-tls-tests-7370
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1400426647000000 1400426647000000
changetime__1445217216921098 1445217216921098
version__None None
owner__exarkun exarkun
cc__felixonmars cc__hynek
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

I can't reproduce any of these failures.  Please provide complete information about the environment you encountered this in.  What platform, what version of Twisted and all of Twisted's dependencies, etc.  Thanks.

**Comment by twisted-trac** [comment]
|<img alt="felixonmars's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|felixonmars commented|
|-|-|

The platform is Arch Linux x86_64 with python 2.7.6 and twisted 14.0.0. Other python dependencies:

pycrypto==2.6.1
zope.interface==4.1.1
service-identity==0.2
pyOpenSSL==0.14
pyasn1==0.1.7
pyasn1-modules==0.0.5

**Comment by twisted-trac** [comment]
|<img alt="felixonmars's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|felixonmars commented|
|-|-|

So after more tests, I found that pyOpenSSL 0.14 seems to be the problem for most of the failures. I've re-run the tests with PyOpenSSL 0.13.1, and only getting 1 failure (attached).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

I don't think `service_identity` is related to these failures.

```
(Twisted)exarkun@top:~/Projects/Twisted/trunk$ svn up -r 41887
...
Updated to revision 41887.
(Twisted)exarkun@top:~/Projects/Twisted/trunk$ bin/trial twisted.test.test_ssl.StolenTCPTestCase.test_properlyCloseFiles
twisted.test.test_ssl
  StolenTCPTestCase
    test_properlyCloseFiles ...                                            [OK]

-------------------------------------------------------------------------------
Ran 1 tests in 0.133s

PASSED (successes=1)
(Twisted)exarkun@top:~/Projects/Twisted/trunk$ svn up -r 41888
...
Updated to revision 41888.
(Twisted)exarkun@top:~/Projects/Twisted/trunk$ bin/trial twisted.test.test_ssl.StolenTCPTestCase.test_properlyCloseFiles
twisted.test.test_ssl
  StolenTCPTestCase
    test_properlyCloseFiles ...                                         [ERROR]

===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/protocols/tls.py", line 358, in _flushReceiveBIO
    bytes = self._tlsConnection.recv(2 ** 15)
  File "/home/exarkun/Environments/Twisted/local/lib/python2.7/site-packages/OpenSSL/SSL.py", line 995, in recv
    self._raise_ssl_error(self._ssl, result)
  File "/home/exarkun/Environments/Twisted/local/lib/python2.7/site-packages/OpenSSL/SSL.py", line 871, in _raise_ssl_error
    _raise_current_error()
  File "/home/exarkun/Environments/Twisted/local/lib/python2.7/site-packages/OpenSSL/_util.py", line 22, in exception_from_error_queue
    raise exceptionType(errors)
OpenSSL.SSL.Error: [('SSL routines', 'SSL23_READ', 'ssl handshake failure')]

twisted.test.test_ssl.StolenTCPTestCase.test_properlyCloseFiles
-------------------------------------------------------------------------------
Ran 1 tests in 0.150s

FAILED (errors=1)
(Twisted)exarkun@top:~/Projects/Twisted/trunk$ 
```

Not visible in this log, I've also verified that the rest of the test failures you reported occur on r41888 but not on r41887.

So it seems r41888 introduced some problems against pyOpenSSL 0.14.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

I wonder if this is related to [#7429](https://github.com/twisted/twisted/issues/7429).

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_9"></a>|@hynek commented|
|-|-|

I’ll focus on test_properlyCloseFiles for now; some observations:

- it works on OS X with OpenSSL 0.9.8 (no ECC)
- it fails on OS X with homebrew’s OpenSSL 1.0.1 (ECC)
- it works on Ubuntu’s 12.04 (ECC in server, but none in client [castrated to TLS 1.1])

Commenting out `self._lib.SSL_CTX_set_tmp_ecdh(context._context, ecKey)` indeed makes the test pass on every platform.  I’ve tried to disable the line only for the server and only for the client but it helped only on the server side.

In panic, I re-checked whether our TLS client-side works at all, and I can confirm that:

- treq can connect to an `twistd web` using HTTPS,
- and check_server_certificate.py can as well.

(I would appreciate a confirmation of these tests by the bug reporters)

----

This leaves us with two alternatives:

1. the testing plumbing is broken for ECC
1. since the test code is using `ClientCreator` instead of SSL4Endpoint, that could be an issue too.  I didn’t have the time to double-check that unfortunately, so it would be great if someone could.

----

Ideas?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

This test is a poster child for the idea of only performing one assertion per test method.

The exception given to the client is terrible because it's the server that decides to screw this test up.  The client doesn't learn anything useful because all it sees is that the server failed the handshake.

The test fails the assertion about the client's connection-lost reason and then we're done.

If we'd gotten to look at the *server*'s exception:

```
Error([('rsa routines', 'RSA_sign', 'digest too big for rsa key'), ('SSL routines', 'SSL3_SEND_SERVER_KEY_EXCHANGE', 'EVP lib')],)
```

I suspect this points somewhat more usefully at the cause of the problem.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [43096]) Branching to 'failing-tls-tests-7370'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @glyph|
|-|-|

Replacing the RSA keys (and therefore also the certificates) used by these tests with larger RSA keys fixed these tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

[#7429](https://github.com/twisted/twisted/issues/7429) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set owner to @exarkun|
|-|-|

Thanks for fixing this, exarkun.

The twistedchecker builder is broken but otherwise the builds look good.  In fact I totally missed the fact where somebody (you, I suspect?) fixed the python 3.3 builders.  Awesome!

I've only got one comment - you should use `getModule(module.__name__).filePath.sibling()` and not `FilePath(module.__file__)` to get module-relative resources.  Assuming that that's functionally identical, please make that change; either way, please land.

(It does occur to me that maybe we ought to be generating our certificate at the beginning of the test run, rather than keeping one persistent one around to eventually become invalid, but that's a discussion for another day…)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

> I've only got one comment - you should use `getModule(module.__name__).filePath.sibling()` and not `FilePath(module.__file__)` to get module-relative resources.

Yep.  I nearly did that - but Python 3.  TLS support is ported but `twisted.python.modules` is not.

> (It does occur to me that maybe we ought to be generating our certificate at the beginning of the test run, rather than keeping one persistent one around to eventually become invalid, but that's a discussion for another day…)

Yes, I find that idea tempting as well but I went with pre-generated for three reasons:

  * smaller change.
  * generating keys is expensive (consuming both wallclock time and system entropy).
  * it's easy to generate a certificate that's valid for the next hundred years, so I did.  (also note the previous certificate expired in 2002 and the tests didn't really care)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_15):
> > I've only got one comment - you should use `getModule(module.__name__).filePath.sibling()` and not `FilePath(module.__file__)` to get module-relative resources.
> 
> Yep.  I nearly did that - but Python 3.  TLS support is ported but `twisted.python.modules` is not.

aauuuggghh.

The only ticket I could find about this was [ticket:6041 this one], so I went ahead and filed [ticket:7644 a new one], mentioning this ticket.

> > (It does occur to me that maybe we ought to be generating our certificate at the beginning of the test run, rather than keeping one persistent one around to eventually become invalid, but that's a discussion for another day…)
> 
> Yes, I find that idea tempting as well but I went with pre-generated for three reasons:
> 
>   * smaller change.

Hence "for another day".

>   * generating keys is expensive (consuming both wallclock time and system entropy).

Speaking of discussions that should happen elsewhere, I'm increasingly convinced that the idea that you can "exhaust system entropy" is just a bug in linux, one which you don't really need to worry about if you use urandom.  You can't exhaust system entropy on any other OS.  But the wall-clock time is a very real problem in tests, so let's burn that bridge when we come to it.

>   * it's easy to generate a certificate that's valid for the next hundred years, so I did.  (also note the previous certificate expired in 2002 and the tests didn't really care)

I did choose my words carefully there - I said "become invalid" and not "expire", since these keys effectively became invalid due to size issues and not due to their expiration date.  Thanks for taking care to make sure the date was appropriately far in the future, though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set status to `closed`|
|-|-|

(In [43114]) Merge failing-tls-tests-7370

Author: exarkun (sponsored by Hybrid Logic, Ltd.)
Reviewer: glyph
Fixes: [#7370](https://github.com/twisted/twisted/issues/7370)

Generate a new and larger RSA key (and associated self-signed certificate)
for use by a number of the TLS unit tests.  This causes them to pass when
the test suite is run using pyOpenSSL 0.14 or newer and an EC-supporting
version of OpenSSL.  The tests previously failed because the newer EC-using
TLS configuration introduced in r41888 requires this larger RSA key (the
old key was 512 bits, an unreasonably small size) to successfully complete
a handshake.

Note this change is unit tested (it is really only a fix to the unit tests)
but there is presently no continuous integration coverage for this case
because the continuous integration system is not configured to run the tests
against pyOpenSSL 0.14 in any cases.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_18"></a>|@hawkowl commented|
|-|-|

Ticket retargeted after milestone deleted
